### PR TITLE
4 playing with hangman dev

### PR DESCRIPTION
**Pull request by shashitekula**

### CODE REVIEW

**Comment by shashitekula** [comment]
Hi Professor,

Please find my assignment 4.

My ID - 46739619

Thanks
Shashi Tekula

File: dictionary/README.md [hunk 1]
```diff
@@ -1,5 +1,9 @@
 
-Please write a couple of sentences explaining your choice of restart
-strategy.
+Choice of restart strategy for Dictionary - 
 
+Dictionary module has Supervisor has "Dictionary.Supervisor" as an children and it has configured supervision strategies as "one-for-one" i.e if a child "Dictionary" process terminates, only that process is restarted.
+	
+the supervisor restarts a child process depending on its :restart configuration.
+
+I have decided to set :restart configuration to :transient, i.e supervisor does not restart the child "Dictionary" in case it exits with reason :normal, :shutdown or {:shutdown, term}.
```
Comment (pragdave) [inline dictionary/README.md:8]: # -2

I see your point, but in practice the Dictionary should never exit normally. In that case, :permanent and :transient are effectively the same, and :permanent better expresses your intent,

File: dictionary/lib/dictionary/dictionary_impl.ex [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  @name Dictionary.Impl
+
+  use Agent
+
+  def start_link() do
+    Agent.start_link( fn -> WordList.word_list  end, name: @name )
+  end
+
+  def random_word() do
+    Agent.get( @name, fn state -> WordList.random_word(state) end )
```
Comment (pragdave) [inline dictionary/lib/dictionary/dictionary_impl.ex:14]: No deduction, just a note that you can also write this as

~~~ elixir
Agent.get(@name, &WordList.random_word/1)
~~~

File: dictionary/test/dictionary_agent_test.exs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+defmodule DictionaryAgentTest do
```
Comment (pragdave) [inline dictionary/test/dictionary_agent_test.exs:1]: # +2

for an extra test!

File: hangman/README.md [hunk 4]
```diff
@@ -1,3 +1,14 @@
...
+Choice of restart strategy for Hangman -
+
+Root supervisor has two children: the Dictionary, and Hangman supervisor.
+The root supervisor uses supervision strategy as "one_for_one" as restart strategy, if a child process terminates, only that process is restarted.
+
+The second-layer supervisor (which supervises the dictionary servers), in turn also uses a one-for-one strategy, so that whenever a it crashes, that dictionary only re-started.
+
+I have decided to set :restart configuration to :permanent, i.e the child process is always restarted.
+
+permanent restart strategy is used, because the Hangman game server must be up and running all times, as it has dependency on other game server workers
```
Comment (pragdave) [inline hangman/README.md:14]: # -4

I really think you want :transient here. If the game crashes, you lose the state, and so there's no real point in restarting it.

**Comment by pragdave** [comment]
Fantastic job!

A small comment on naming: `Hangman.HangmanImpl` is normally just `Hangman.Impl`.

**Pull request closed without merging**